Using ASP.NET HtmlEncoder instance

sebastienros opened the pull request:
Fixes #4041

sebastienros commented:
Asking for reviews as I changed the way some things were encoded where it didn't make sense to me.

jtkech commented on src/OrchardCore.Modules/OrchardCore.Markdown/Drivers/MarkdownBodyPartDisplay.cs:
```diff
@@ -73,9 +76,8 @@ private async ValueTask BuildViewModel(MarkdownBodyPartViewModel model, Markdown
-
-            model.Html = Markdig.Markdown.ToHtml(markdown ?? "");
+            model.Markdown = await _liquidTemplatemanager.RenderAsync(MarkdownBodyPart.Markdown, NullEncoder.Default, templateContext);
+            model.Html = Markdig.Markdown.ToHtml(model.Markdown ?? "");
```
Looks right to use the `NullEncoder` to initialize the `Markdown` field

But then, to set the `Html` field maybe we need

    model.Html = _htmlEncoder.Encode(Markdig.Markdown.ToHtml(model.Markdown ?? "");

Here, assuming that `Markdig.Markdown.ToHtml()` doesn't do any html encoding.
And knowing that the related shape do `@Html.Raw(Model.Html)` <= but maybe wrong here

Note: For the html body part we use `_htmlEncoder` and the shape also uses `@Html.Raw(Model.Html)`.

sebastienros replied:
Liquid should use the html encoder here. And we should not encode the markdown.

jtkech commented on src/OrchardCore.Modules/OrchardCore.Markdown/Drivers/MarkdownFieldDriver.cs:
```diff
@@ -38,7 +41,7 @@ public override IDisplayResult Display(MarkdownField field, BuildFieldDisplayCon
 
-                model.Markdown = await _liquidTemplatemanager.RenderAsync(field.Markdown, System.Text.Encodings.Web.HtmlEncoder.Default, templateContext);
+                model.Markdown = await _liquidTemplatemanager.RenderAsync(field.Markdown, NullEncoder.Default, templateContext);
                 model.Html = Markdig.Markdown.ToHtml(model.Markdown ?? "");
```
Idem, looks right to use the `NullEncoder` to initialize the `Markdown` field

But then, to set the `Html` field maybe we need

    model.Html = _htmlEncoder.Encode(Markdig.Markdown.ToHtml(model.Markdown ?? "");

Knowing that the related shape do `@Html.Raw(Model.Html)` <= but maybe wrong here

Note: For the html field shape we also uses `@Html.Raw(Model.Html)` and the driver doesn't use `_htmlEncoder`, so maybe we also need to update the html field driver.

sebastienros replied:
Liquid should use the html encoder here. And we should not encode the markdown.

jtkech commented on src/OrchardCore.Modules/OrchardCore.Markdown/GraphQL/MarkdownBodyQueryObjectType.cs:
```diff
@@ -33,7 +34,7 @@ private static async Task<object> ToHtml(ResolveFieldContext<MarkdownBodyPart> c
+            markdown = await liquidTemplateManager.RenderAsync(markdown, NullEncoder.Default, templateContext);
 
             return Markdig.Markdown.ToHtml(markdown);
         }
```
Idem, knowing that the `HtmlBodyQueryObjectType` uses the `htmlEncoder`.

sebastienros replied:
My mistake, now I realize that the liquid encoding is only for injected values, so this should definitely be encoded at this level. I will fix it where I changed it.

jtkech commented on src/OrchardCore.Modules/OrchardCore.Markdown/GraphQL/MarkdownFieldQueryObjectType.cs:
```diff
@@ -33,7 +34,7 @@ private static async Task<object> ToHtml(ResolveFieldContext<MarkdownField> ctx)
-            markdown = await liquidTemplateManager.RenderAsync(markdown, templateContext);
+            markdown = await liquidTemplateManager.RenderAsync(markdown, NullEncoder.Default, templateContext);
 
             return Markdig.Markdown.ToHtml(markdown);
```
Idem

jtkech commented:
- I merged dev and fixed CI Tests

- Then i did some minor changes, see my last commit for the little suggested changes.

- Then i did a little review but TBH i'm lost about when we need to encode or not, i need to re-think about it tomorrow ;)

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.Liquid/Filters/LiquidFilter.cs:
```diff
@@ -25,7 +28,7 @@ public async ValueTask<FluidValue> ProcessAsync(FluidValue input, FilterArgument
 
-            var content = await _liquidTemplateManager.RenderAsync(input.ToStringValue(), context);
+            var content = await _liquidTemplateManager.RenderAsync(input.ToStringValue(), _htmlEncoder, context);
 
```
Basically agree with what @jtkech has said, and also confused about what should be encoded vs what shouldn't.

Think it makes sense to move all the liquid rendering to use the injected `HtmlEncoder`, and get rid of the extension method override that was being used here. 

The `Html.Raw()` in the Html template is a bit questionable though, not sure.
I would ask the same question about the agency templates where I used 
`{{ Model.ContentItem.Content.RawHtml.Content.Html | liquid | raw }}` and was not entirely sure that it needed the `raw` pipe (and I think you questioned that in the meeting when going through PR's)

sebastienros replied:
If the property contains Html then you should use ` | raw`.

sebastienros replied:
Otherwise you'll get all tags encoded and be rendered as is, kind of obvious while I am stating that.

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.Html/Handlers/HtmlBodyPartHandler.cs:
```diff
@@ -45,7 +46,7 @@ public override Task GetContentItemAspectAsync(ContentItemAspectContext context,
-                    var result = await _liquidTemplateManager.RenderAsync(part.Html, HtmlEncoder.Default, templateContext);
+                    var result = await _liquidTemplateManager.RenderAsync(part.Html, _htmlEncoder, templateContext);
                     bodyAspect.Body = _bodyAspect = new HtmlString(result);
                 }
```
How will this work (noting that it's no different to what we do already) with the Lucene full text aspect?

which uses this `bodyAspect.Body.WriteTo(sw, NullHtmlEncoder.Default);` but chances are good the `BodyAspect` will be coming through encoded by rendering any liquid in it?

sebastienros replied:
`bodyAspect.Body.WriteTo(sw, NullHtmlEncoder.Default);` means that it won't encode the html, so we'll keep the html as is.

sebastienros commented on src/OrchardCore/OrchardCore.DisplayManagement.Liquid/HtmlContentValue.cs:
```diff
@@ -57,7 +58,7 @@ public override string ToStringValue()
         public override void WriteTo(TextWriter writer, TextEncoder encoder, CultureInfo cultureInfo)
         {
-            _value.WriteTo(writer, HtmlEncoder.Default);
+            _value.WriteTo(writer, NullHtmlEncoder.Default);
```
We need to double-check this one, where is this HtmlContentValue used.

jtkech replied:
I will check this one.

jtkech replied:
@sebastienros

So, it is only used for liquid in the `ShapeStringify` and `ShapeRender` filters that use the `displayHelper` to execute shapes. It is a wrapper holding an `IHtmlContent` that may be a simple `HtmlString` or e.g a `ViewBuffer` composed of `string` / `HtmlString` and others `IHtmlContent` that need to be encoded or not depending on their implementation.

So, i tried the following with some special charaters in an article Subtitle.

    {{ Model.Content["Article-Subtitle"] | shape_render }}

This with the `HtmlEncoder` or the `NullHtmlEncoder`

And mixed with these 2 cases in `TextField.cshtml`

    @Model.Text
    @(new StringHtmlContent(Model.Text))

With `@Model.Text` the result is the same with or without the `HtmlEncoder`. We retrieve it above as an already encoded string. Note: with `@Html.Raw()` we retrieve it as an `HtmlString` => idem same results but not encoded.

    @Model.Text
    _value.WriteTo(writer, (HtmlEncoder)encoder);
     This is what I do. Special characters &gt; &#xE9;&#xE9;&#xE9;

    @Model.Text
     _value.WriteTo(writer, NullHtmlEncoder.Default);
    This is what I do. Special characters &gt; &#xE9;&#xE9;&#xE9;

But with `@(new StringHtmlContent(Model.Text))` the results are different.

    @(new StringHtmlContent(Model.Text))
    _value.WriteTo(writer, NullHtmlEncoder.Default);
    This is what I do. Special characters > ééé

    @(new StringHtmlContent(Model.Text))
    _value.WriteTo(writer, (HtmlEncoder)encoder);
    This is what I do. Special characters &gt; &#xE9;&#xE9;&#xE9;

My conclusion would be that it is better to use `(HtmlEncoder)encoder`.

sebastienros commented on src/OrchardCore/OrchardCore.DisplayManagement/PositionWrapper.cs:
```diff
@@ -30,7 +30,7 @@ public PositionWrapper(IHtmlContent value, string position)
         public PositionWrapper(string value, string position)
         {
-            _value = new HtmlString(HtmlEncoder.Default.Encode(value));
+            _value = new HtmlString(value);
```
I need to double-check this one, I think it might still need to get encoded.

jtkech replied:
I will check this one

jtkech replied:
It is only used in `Shape.cs`, we use it when adding an item that is a `string` or an `IHtmlContent`. If it is a string we convert it to an `HtmlString`, because _value is an `IHtmlContent`, and that's why we were encoding the original string.

Yes, i think it needs to be encoded.

But maybe better to use here

    using Microsoft.AspNetCore.Mvc.ViewFeatures
    ...

    _value = new StringHtmlContent(value);

So that it will be encoded through the `PositionWrapper.WriteTo()` method and with the right encoder.

sebastienros commented:
Just to summarize the rules I have applied:
- If something was using `HtmlEncoder.Default` I replaced it with the DI one
- I removed the extension method for Liquid that didn't take any encoder, now the caller is responsible for resolving the encoder and using it

jtkech commented:
I'll try to check the ones you mentioned.

I'm working on it by testing things step by step, i made good progress but still a little lost ;)

**Here some 1st findings**

- Yes, not using `@Html.Raw(Model.Html)` would end up to encode the html tags themselves. Razor encodes thing when you use the `@` character referencing c# code, but not if it is an `HtmlString`.

- `_liquidTemplatemanager.RenderAsync()` using `_htmlEncoder` only encode the parts using liquid syntax. Here characters as `>` and also `é` are encoded by the current encoder. Seems to be consistent with what razor does when using the `@` character.

- So at this point, i'm thinking that outside a Razor / Liquid expression, this is the author (or an editor tool) that needs to provide the desired html entities.

- `Markdown.ToHtml()` does some encoding and seems to only encode html reserved characters e.g `>`, but not e.g `é`, no way to use the current encoder. **Moreover, if it finds an encoded character which is not a reserved char e.g `#xE9;` it reverts it to `é`**. So, before `ToHtml()`, using `RenderAsync()` with an non null encoder or not seems to have the same result.

    ![Encoding](https://user-images.githubusercontent.com/8586360/69196376-9ede6e80-0b2e-11ea-8ac3-4ef6cb4f662a.png)

    I would suggest to 1st do `ToHtml()` and then `RenderAsync()` with the current encoder. Here also `ToHtml()` reverts back the hardcoded html entities to their original character (if not an html reserved one), but we have a better chance to encode chars included in a liquid expression (not hardcoded).

jtkech commented on src/OrchardCore.Modules/OrchardCore.Markdown/Drivers/MarkdownBodyPartDisplay.cs:
```diff
@@ -73,7 +76,7 @@ private async ValueTask BuildViewModel(MarkdownBodyPartViewModel model, Markdown
 
-            var markdown = await _liquidTemplatemanager.RenderAsync(MarkdownBodyPart.Markdown, System.Text.Encodings.Web.HtmlEncoder.Default, templateContext);
+            var markdown = await _liquidTemplatemanager.RenderAsync(MarkdownBodyPart.Markdown, _htmlEncoder, templateContext);
 
```
Just to say that here we use a variable `var markdown =` and in the field driver we update the model `model.Markdown =`. Not sure what's the best but maybe better to do the same in both drivers.

A rule would be to set as many things we can to the template context, and then, after having evaluated the liquid expressions, update the model as much as we can.

Sorry to bother you with a so small thing ;) Anyway that's ok for me.

jtkech commented:
@sebastienros i will do a little commit related to my last review comments so that you will see what i meant and then agree or revert them .

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Markdown/Drivers/MarkdownFieldDriver.cs:
```diff
@@ -41,8 +41,9 @@ public override IDisplayResult Display(MarkdownField field, BuildFieldDisplayCon
 
-                model.Markdown = await _liquidTemplatemanager.RenderAsync(field.Markdown, _htmlEncoder, templateContext);
-                model.Html = Markdig.Markdown.ToHtml(model.Markdown ?? "");
+                var markdown = await _liquidTemplatemanager.RenderAsync(field.Markdown, _htmlEncoder, templateContext);
```
I instead want to change the other one to have only Markdown (no liquid) in the Markdown property. We also do this in the HtmlField and HtmlBodyPart to store the Html without liquid.

jtkech replied:
Okay

sebastienros commented:
I'd love to find more locations where we can use `StringHtmlContent`, will do a quick look

jtkech commented:
Okay cool

sebastienros commented:
Updated

jtkech commented:
I will take a look

jtkech commented:
LGTM !

ampadashi commented:
I've tried this version, but it still has some problems. When we use Email Task in a workflow, If we use Unicode character in the body of the email,  it will not be render properly:
![image](https://user-images.githubusercontent.com/32009041/69628768-18abb600-1061-11ea-9fa6-bce1284fba75.png)

ampadashi commented:
> I've tried this version, but it still has some problems. When we use Email Task in a workflow, If we use Unicode characters in the body of the email, they will not be rendered properly:
> ![image](https://user-images.githubusercontent.com/32009041/69628768-18abb600-1061-11ea-9fa6-bce1284fba75.png)
> It's so wired. the texts that are get from the textarea rendered in the email properly. But, those ones which are get from the input[type=text] will not be rendered properly.

I've just added raw filter to the liquid models used in the email template and the problem solved.
